### ISSUE TITLE
Is LiteCollection<T>.Update(IEnumerable<T> documents) atomic in v4?

### ISSUE THREAD

**Issue by resnickj** [litedb-org/LiteDB#2053]
I am trying to figure out if the `LiteCollection<T>.Update(IEnumerable<T> documents) `method behaves atomically (all updates are performed, or none are) in v4 ? Looking in the code I see an internal Transaction being used here, which suggests the answer is yes, but then in the documentation I read that transactions were removed in v4, so I'm confused.

**Comment by JKamsker** [comment]
Yes. In v4, a single `Update(IEnumerable<T>)` call is all-or-nothing. What v4 removed was the user-facing `BeginTrans()` API for grouping several calls; each single engine call still runs in an internal transaction. In 4.1.4:

- The whole loop over your documents runs under one exclusive database lock (`LiteEngine.Transaction<T>`).
- Changed pages stay in memory until the loop finishes and are then written once (`PersistDirtyPages`). The `CheckPoint()` inside the loop only drops unchanged pages from the cache; it never writes anything.
- If anything throws, including your own code while the input sequence is enumerated, all changed pages are discarded and nothing is written.
- With `Journal=true` (the default), the pages go to the journal file first, so a crash during the final write is recovered on the next open. With `Journal=false`, a crash at that moment can leave the file partially written.

One thing to watch: very large batches keep all changed pages in memory until the end.

v4 is no longer developed. In v5 the same call also runs in one transaction, and explicit `BeginTrans()`/`Commit()` is back.

Closing this as answered — if it doesn't cover your case, comment here and we'll reopen it.
